services/shhext/history.go

```diff
@@ -114,10 +137,69 @@ func (reactor *HistoryUpdateReactor) CreateRequests(topicRequests []TopicRequest
 			filtered = append(filtered, req)
 		}
 	}
-	filtered = append(filtered, GroupHistoriesByRequestTimespan(reactor.store, mapToList(histories))...)
+	adjusted, err := adjustRequestedRanges(reactor.store, mapToList(histories))
+	if err != nil {
+		return nil, err
+	}
+	filtered = append(filtered,
+		GroupHistoriesByRequestTimespan(reactor.store, adjusted)...)
 	return RenewRequests(filtered, reactor.timeSource()), nil
 }
 
+// for every history that is not included in any request check if there are other ranges with such topic in db
+// if so check if they can be merged
+// if not adjust second part so that End of it will be equal to First of anothe
+func adjustRequestedRanges(store db.HistoryStore, histories []db.TopicHistory) ([]db.TopicHistory, error) {
```

Comment:
can you add a diagram what does it do exactly?...

or maybe a separate unit-test for that function to actually show how it is used, because this algorithm is a bit hard to understand out of context.

also, tests for this particular function should be state-less and help with improving the code further

Reply:
there are tests for every case in this method. all of them start with TestAdjustHistories